### PR TITLE
feat: add script to recover missing data in archival nodes

### PR DESCRIPTION
**Pull request by Trisfald**
This script contains the minimal set of commands to restore missing data from archival nodes.

The data loss appeared in three time periods:
- early January
- after first resharding
- after second resharding

Due to the nature of how trie nodes are stored in the database, one missing node in the past may cause errors when querying accounts at very distant future heights.

Recovery takes many days and required an SSD backed cold storage. 

I've tested the procedure's outcome with manual queries and by iterating tries from different state roots.

### CODE REVIEW

**Review by wacban — COMMENTED** [review]
Nice, it looks really clean! 

What is the goal for this script? Do you want to share it with the node operators to run in order to recover the missing data? I wonder if a doc with a description of the problem and the commands would be better for this purpose? What do you think? 

Can you also describe how to check if the data is missing and that it was recovered?

File: scripts/recover_missing_archival_data.sh [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+#
+# On top of every command, where applicable, we include the number of expected database writes.
+# The estimated number of batches assumes a batch size of 500.0 KB. 
+#
+# The entire procedure is expected to take several days to complete.
+
+##########
+# neard must be stopped
+##########
+systemctl is-active --quiet neard && echo "neard must be stopped" && exit 1
```
Comment (wacban) [inline scripts/recover_missing_archival_data.sh:22]: This is how we run neard in nearone, it doesn't necessarily need to be the same for other node operators.

**Review by wacban — COMMENTED** [review]
It may be worth displaying some high level logs from the scripts itself too. In case it's interrupted the user would want to resume rather than restart.

**Comment by Trisfald** [comment]
> Nice, it looks really clean!
> 
> What is the goal for this script? Do you want to share it with the node operators to run in order to recover the missing data? I wonder if a doc with a description of the problem and the commands would be better for this purpose? What do you think?
> 
> Can you also describe how to check if the data is missing and that it was recovered?

Honestly, I wouldn't recommend running this script to node operators. I hope we can distribute a fully recovered DB snapshot and keep this as fallback strategy.
The reason is that even with best hardware just running the recovery and then catching up to be in sync again will take ~1 month

File: scripts/recover_missing_archival_data.sh (hunk 1 above)
Comment (Trisfald) [reply scripts/recover_missing_archival_data.sh:22]: Good point

it's just for convenience - if you run the commands with a running `neard` you end up getting `error DB already locked`

**Comment by Trisfald** [comment]
> It may be worth displaying some high level logs from the scripts itself too. In case it's interrupted the user would want to resume rather than restart.

yes I put `-x` option for minimal traceability

File: scripts/recover_missing_archival_data.sh [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+
+##########
+# Recover missing trie nodes before resharding by re-applying blocks
+##########
+export RUST_LOG=info
+
+# DB writes = 377624
```
Comment (tayfunelmas) [inline scripts/recover_missing_archival_data.sh:30]: what is this? expected output from the next command?

File: scripts/recover_missing_archival_data.sh [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+/home/ubuntu/neard view-state -t cold --readwrite apply-range --start-index 109913254 --end-index 110050000 --shard-id 2 --storage trie-free --save-state cold sequential
+
+
+##########
+# Perform first resharding
+##########
+export RUST_LOG=debug
+
+# DB writes >= 133646770, batch_count = 16569
+/home/ubuntu/neard database resharding --height 114580307 --shard-id 0 --restore
```
Comment (tayfunelmas) [inline scripts/recover_missing_archival_data.sh:40]: I would echo some message before each of these eg., "Resharding database at height ... and shard ..."

File: scripts/recover_missing_archival_data.sh (hunk 2 above)
Comment (tayfunelmas) [reply scripts/recover_missing_archival_data.sh:30]: ok, the comment already talks about this.

File: scripts/recover_missing_archival_data.sh [hunk 4]
```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+set -eox pipefail
+
+# This script can be used on archival nodes to recover missing data from early 2024.
+#
+# Requirements:
+# - The cold store must be mounted on an SSD to make the operation faster.
+# - The config 'resharding_config.batch_delay' must be set to 0.
```
[the remark this answers is not in the document]
Comment (Trisfald) [reply scripts/recover_missing_archival_data.sh:8]: actually the check is done inside `neard` resharding tool

**Review by wacban — APPROVED** [review]
LGTM

**Pull request merged**